### ISSUE TITLE
Authentication Hook

### ISSUE THREAD

**Issue by behrad** [mcollina/aedes#9]
Another issue with Mosca is the race with plugging authentication hook, and that is because of setting auth handler after server boot, not as an configuration option.

```
server = new mosca.Server(moscaSettings);
//server is now ready and may accept new connections
server.authenticate = authorizer.authenticate;
server.authorizeSubscribe = authorizer.authorizeSubscribe;
server.authorizePublish = authorizer.authorizePublish;
```

This will be important on server restarts in production (with heavy connected clients)

**Comment by mcollina** [comment]
This won't be true here, as Aedes does not handle the connections.
However I agree we should support setting there in the config.

Would you like to send a PR?

**Comment by behrad** [comment]
How aedes will handle auth? a plugin-like architecture or a handler function in config?
What is the specification (methods) for such handler?

**Comment by mcollina** [comment]
Similar to Mosca: https://github.com/mcollina/aedes#instanceauthenticateclient-username-password-doneerr-successful.

**Comment by mcollina** [comment]
You can override some methods, but I think it might be easier to just provide those functions in the options, rather than override the methods.

**Comment by marvinroger** [comment]
So basically, add `authenticate`, `authorizePublish` and `authorizeSubscribe` to `aedes()` options?

**Comment by mcollina** [comment]
@marvinroger that would be the desired behavior, I'll accept a PR on that if you would like to contribute. Currently you have to override.

**Comment by pkinney** [comment]
Made a PR (#36) to attempt to address this.

**Comment by mcollina** [comment]
Released as 0.13.0.
